### ISSUE TITLE
Caching fails when default argument value is nan

### ISSUE THREAD

**Issue by polakowo** [numba/numba#8664]
## Reporting a bug

- [x] I have tried using the latest released version of Numba (most recent is
 visible in the change log (https://github.com/numba/numba/blob/main/CHANGE_LOG).
- [x] I have included a self contained code sample to reproduce the problem.
  i.e. it's possible to run as 'python bug.py'.

File test.py:

```python
from numba import njit

@njit(cache=True)
def f_finite(x=0):
    pass

@njit(cache=True)
def f_nan(x=float("nan")):
    pass

@njit(cache=True)
def f_inf(x=float("inf")):
    pass
```

Running the functions for the _second_ time:

```python
import os
os.environ['NUMBA_DEBUG_CACHE'] = "1"

from test import f_finite, f_nan, f_inf

f_finite()
[cache] index loaded from '/.../test.f_finite-3.py39.nbi'
[cache] data loaded from '/.../test.f_finite-3.py39.1.nbc'

f_nan()
[cache] index loaded from '/.../test.f_nan-7.py39.nbi'
[cache] index loaded from '/.../test.f_nan-7.py39.nbi'
[cache] index saved to '/.../test.f_nan-7.py39.nbi'
[cache] data saved to '/.../test.f_nan-7.py39.2.nbc'

f_inf()
[cache] index loaded from '/.../test.f_inf-11.py39.nbi'
[cache] data loaded from '/.../test.f_inf-11.py39.1.nbc'
```

While I understand that it's all based on identity checks and nan != nan and one can use None as a substitution, every time the users passes `np.nan` (for example as a fill value for an array), the function will be needlessly re-cached.

**Comment by esc** [comment]
@polakowo thank you for submitting this. I can confirm I am able to reproduce this with the current `main` at 2f2e2c1a7f940977e37901f4289b771fedbe4af2 --yes I believe that the fact that nan is always unequal to nan is the cause here. I suppose when comparing a function, it may be possible to override this default.

**Comment by sklam** [comment]
I don't think this is something that we should fix. It is similar to the best practice of not using NAN as key of dictionary. 

Python `inspect` also treat two instances of the signature of `f_nan` as not equal:

```python
import inspect

def f_nan(x=float("nan")):
    pass

def f_inf(x=float("inf")):
    pass


assert inspect.signature(f_inf) == inspect.signature(f_inf)   # this is ok
assert inspect.signature(f_nan) == inspect.signature(f_nan)   # this is not
```

**Comment by polakowo** [comment]
NaN not being equal to NaN is an important principle in scientific computing, but I don't see any advantage of having NaN being treated differently across function signatures since two identical functions taking NaN are guaranteed to produce the same compiled code. It took me a while to understand why a bunch of my functions produce tons of cache without any particular reason (on the other hand, a zero-dimensional array with NaN will do fine because of the same data type). 

I also wouldn't want to rely on None since explicitly passing NaN would still re-cache the entire thing each time.

**Comment by stuartarchibald** [comment]
I wonder if a workaround might be to create a custom float type that overrides `__eq__` and `__hash__` to do something special for the `nan` value and then use that?

**Comment by sklam** [comment]
a `HashableNan` just for Numba?

**Comment by sklam** [comment]
> I also wouldn't want to rely on None since explicitly passing NaN would still re-cache the entire thing each time.

If the default value is NaN, you will get two versions:

1. x: `omitted(default=NaN)` --- goes here when the argument is omitted.
2. x: `float64`  --- explicitly passing NaN or other float value goes here.

The `None` case will get 3 versions:

1. x: `omitted(default=None)` --- omitting arg
2. x: `None`  --- explicitly passing `None`
3. x: `float64` --- passing in float

**Comment by polakowo** [comment]
@sklam you're right, passing NaN is reusing cache as expected.
